### ISSUE TITLE
app.prepare() failure crash the program but error message hidden

### ISSUE THREAD

**Issue by Josehower** [vercel/next.js#47211]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
➜  courses git:(express-next-gatsby-mvp) ✗ npx --no-install next info

    Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Fri Jan 27 02:56:13 UTC 2023
    Binaries:
      Node: 18.14.2
      npm: 9.5.0
      Yarn: 1.22.19
      pnpm: 7.18.2
    Relevant packages:
      next: 13.2.4
      eslint-config-next: 13.2.4
      react: 18.2.0
      react-dom: 18.2.0

warn  - Latest canary version not detected, detected: "13.2.4", newest: "13.2.5-canary.4".
        Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
        Read more - https://nextjs.org/docs/messages/opening-an-issue
```

### Link to the code that reproduces this issue

https://stackblitz.com/edit/vercel-next-js-z4hyvr?file=README.md,server.js,package.json,app%2Fpage.tsx,next.config.js,server-try-catch.js

### To Reproduce

Run the project with scripts `"dev": "node server.js"` and `"dev-try-catch": "node server-try-catch.js"`

Theoretically, both custom servers should crash in the same way but don't

### Describe the Bug

When using await on `app.prepare()`, the server doesn't start, but also, and the error message is hidden, which is very confusing about the reasons it is crashing. to make the output error clear, the `app.prepare()` must be wrapped in a try-catch block using console.log().

This is inconvenient since both await and all try catch versions should behave similarly and print an error.

using only await, `throw err` and `throw new Error(err)`  don't print the error message

```js
  try {
    await app.prepare();
  } catch (err) {
    throw err // throw new Error(err) also not work
  }
```

```js
await app.prepare();
```

output:

```bash
~/projects/vercel-next-js-z4hyvr 12s❯ npm run dev
...

success Saved lockfile "package-lock.json"
success Updated "package.json"

success Install finished in 0.059s
```
  
Only when the `try-catch` contains a `console.error()` the error message is printed
  
```js
  try {
    await app.prepare();
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
```
output:

```bash
~/projects/vercel-next-js-z4hyvr 12s ❯ npm run dev-try-catch
...

success Saved lockfile "package-lock.json"
success Updated "package.json"

success Install finished in 0.062s
$ node server-try-catch.js
Error: > The `app` directory is experimental. To enable, add `appDir: true` to your `next.config.js` configuration under `experimental`. See https://nextjs.org/docs/messages/experimental-app-dir-config
    at Object.findPagesDir (file:///home/projects/vercel-next-js-z4hyvr/node_modules/next/dist/lib/find-pages-dir.js:80:19)
    at DevServer.getRoutes (file:///home/projects/vercel-next-js-z4hyvr/node_modules/next/dist/server/dev/next-dev-server.js:138:59)
    at new Server (file:///home/projects/vercel-next-js-z4hyvr/node_modules/next/dist/server/base-server.js:108:47)
    at new NextNodeServer (file:///home/projects/vercel-next-js-z4hyvr/node_modules/next/dist/server/next-server.js:70:9)
    at new DevServer (file:///home/projects/vercel-next-js-z4hyvr/node_modules/next/dist/server/dev/next-dev-server.js:97:9)
    at NextServer.createServer (file:///home/projects/vercel-next-js-z4hyvr/node_modules/next/dist/server/next.js:144:20)
    at eval (file:///home/projects/vercel-next-js-z4hyvr/node_modules/next/dist/server/next.js:159:42)
    at async NextServer.prepare (file:///home/projects/vercel-next-js-z4hyvr/node_modules/next/dist/server/next.js:134:24)
    at async startNext (file:///home/projects/vercel-next-js-z4hyvr/server-try-catch.js:14:5)

```

### Expected Behavior

`await app.prepare();` should crash with a proper error message.

try-catch blocks using  `throw err` and `throw new Error(err)`  should print an error message

**Comment by karlhorky** [comment]
Looks like this may be related to the `uncaughtException` and `unhandledRejection` handlers and `logError` function in `initialize()` in `src/server/lib/router-server.ts`:

https://github.com/vercel/next.js/blob/01e5185583970e0a068e8b322a537f2a7368d9f9/packages/next/src/server/lib/router-server.ts#L628-L641

This will cause the following error to be swallowed and the server to stop with no message and exit code 0, which seems like very unusual behavior:

`server.js`

```js
import next from 'next';

const app = next({
  dev: process.env.NODE_ENV !== 'production',
  hostname: 'localhost',
  port: 3000,
});

await app.prepare();
throw new Error('I am swallowed');
```

```bash
$ node server.js
# no output
$ echo $?
0
```

Apparently the community has asked for a way to log exceptions / errors on custom servers:

- https://github.com/vercel/next.js/issues/4808
- https://github.com/vercel/next.js/discussions/32215

The current solution proposed by @leerob is to use OpenTelemetry for this:

- https://github.com/vercel/next.js/discussions/32215#discussioncomment-7426703
- https://nextjs.org/docs/app/building-your-application/optimizing/open-telemetry

Not sure exactly how it relates though to logging errors.

Or maybe this a bug? cc @ijjk @shuding

**Comment by karlhorky** [comment]
## Workaround

Override `uncaughtException` (and possibly also `unhandledRejection`) in non-development environments:

`server.js`

```js
import next from 'next';
import { isPostpone } from 'next/dist/server/lib/router-utils/is-postpone.js';

const app = next({
  dev: process.env.NODE_ENV !== 'production',
  hostname: 'localhost',
  port: 3000,
});

await app.prepare();

// Overwrite uncaughtException handler in production to throw errors
// https://github.com/vercel/next.js/issues/47211
if (!isDevelopment) {
  // Copied from src/server/lib/router-server.ts
  // https://github.com/vercel/next.js/blob/01e5185583970e0a068e8b322a537f2a7368d9f9/packages/next/src/server/lib/router-server.ts#L628-L638
  process.on('uncaughtException', (error: Error | undefined) => {
    // React postpones that are unhandled might end up logged here but they're
    // not really errors. They're just part of rendering.
    if (isPostpone(error)) return;

    throw error;
  });
}

throw new Error('I am no longer swallowed');
```

Output:

```bash
$ node server.js
file:///Users/k/p/project/server.js:31
        throw error;
        ^

Error: I am no longer swallowed
    at file:///Users/k/p/project/server.js:36:7

Node.js v20.15.0
```

**Comment by marcoshernanz** [comment]
Thanks for the report. I reproduced the silent `app.prepare()` failure on 13.2.4, but it stopped reproducing in 13.4.0 after #48365. On current stable and canary, failed preparation prints the full stack and exits nonzero, and errors thrown after preparation are also logged. The remaining nonfatal process behavior is intentional and separate from the hidden-error bug, so I’m closing this as fixed.
